File: src/mpc.cpp
#include "ros_controllers/mpc.hpp"



ROS2Controllers::MPCController::MPCController(double dt, int horizon, double L, std::vector<double> Q_param, std::vector<double> R_param, double signal_limit_linear_velocity, 
    double signal_limit_angular_velocity, double error_threshold)
    : dt_(dt), N_(horizon), L_(L), Q_vector_(Q_param), R_vector_(R_param), error_threshold_(error_threshold), 
      signal_limit_linear_velocity_(signal_limit_linear_velocity), signal_limit_angular_velocity_(signal_limit_angular_velocity), 
      discrete_linear_error_(0.0), continous_linear_error_(0.0) {

    SX x = SX::sym("x");
    SX y = SX::sym("y");
    SX theta = SX::sym("theta");
    states_ = SX::vertcat({x, y, theta});
    n_states_ = states_.size1();
    
    SX v = SX::sym("v");
    SX omega = SX::sym("omega");
    controls_ = SX::vertcat({v, omega});
    n_controls_ = controls_.size1();

    // System dynamics (unicycle model)
    SX rhs = SX::vertcat({
        v * SX::cos(theta),
        v * SX::sin(theta),
        omega
    });

    // Convert the system dynamics to a function
    f_ = Function("f", {states_, controls_}, {rhs});

    // MPC variables
    U_ = SX::sym("U", n_controls_, N_);          
    X_ = SX::sym("X", n_states_, N_ + 1);
    P_ = SX::sym("P", n_states_ + N_ * n_states_);

    // Objective function and constraints
    SX obj = 0;
    std::vector<SX> g;

    // Initial state constraint
    SX st = X_(Slice(), 0);
    g.push_back(st - P_(Slice(0, n_states_)));

    // Cost function (state and control weights)
    SX Q = SX::diagcat(std::vector<SX>{SX(Q_vector_[0]), SX(Q_vector_[1]), SX(Q_vector_[2])});
    SX R = SX::diagcat(std::vector<SX>{SX(R_vector_[0]), SX(R_vector_[1])});

    // MPC loop
    for (int k = 0; k < N_; ++k) {
        st = X_(Slice(), k);
        SX con = U_(Slice(), k);
        SX st_next = X_(Slice(), k + 1);
        SX f_value = f_(std::vector<SX>{st, con})[0];
        SX st_next_euler = st + dt_ * f_value;
        g.push_back(st_next - st_next_euler);

        // Reference state
        SX ref = P_(Slice(n_states_ + k * n_states_, n_states_ + (k + 1) * n_states_));

        // Update the objective function
        obj = obj + mtimes((st - ref).T(), mtimes(Q, st - ref)) + mtimes(con.T(), mtimes(R, con));
    }

    // Create the concatenated constraint vector
    SX OPT_variables = SX::vertcat(std::vector<SX>{
        SX::reshape(U_, n_controls_ * N_, 1),
        SX::reshape(X_, n_states_ * (N_ + 1), 1)
    });

    SX G = SX::vertcat(g);

    // Create the NLP problem
    casadi::SXDict nlp_prob = {{"f", obj}, {"x", OPT_variables}, {"g", G}, {"p", P_}};

    // Options for the solver
    Dict opts;
    opts["ipopt.max_iter"] = 2000;
    opts["ipopt.print_level"] = 0;
    opts["print_time"] = 0;
    opts["ipopt.acceptable_tol"] = 1e-8;
    opts["ipopt.acceptable_obj_change_tol"] = 1e-6;

    // Create the solver function
    solver_ = nlpsol("solver", "ipopt", nlp_prob, opts);

    // Initialize control and states values for predictions
    u0_ = DM::zeros(N_, n_controls_);
    x_mpc_ = DM::zeros(N_ + 1, n_states_);

    // State and control constraints
    int n_vars = OPT_variables.size1();
    lbx_ = std::vector<double>(n_vars, -std::numeric_limits<double>::infinity());
    ubx_ = std::vector<double>(n_vars, std::numeric_limits<double>::infinity());

    double v_max = signal_limit_linear_velocity;
    double v_min = -signal_limit_linear_velocity;
    double omega_max = signal_limit_angular_velocity;
    double omega_min = -signal_limit_angular_velocity;

    for (int k = 0; k < N_; ++k) {
        lbx_[k * 2] = v_min;
        ubx_[k * 2] = v_max;

        lbx_[k * 2 + 1] = omega_min;
        ubx_[k * 2 + 1] = omega_max;
    }

    // Current state
    x0_ = DM::zeros(n_states_);

    ref_traj_ = DM::zeros(N_, n_states_);
}



ROS2Controllers::MPCController::~MPCController() {

}



void ROS2Controllers::MPCController::testFunction() {

}


double ROS2Controllers::MPCController::getDiscreteLinearError() {
    return discrete_linear_error_;
}



double ROS2Controllers::MPCController::getContinousLinearError() {
    return continous_linear_error_;
}



void ROS2Controllers::MPCController::setupOptimizationProblem() {

}



void ROS2Controllers::MPCController::updateReferenceTrajectory(const std::vector<Eigen::VectorXd>& reference_trajectory) {
    /*
    for (int k = 0; k < N_; ++k) {
        ref_traj_(k, 0) = reference_trajectory[k](0);
        ref_traj_(k, 1) = reference_trajectory[k](1);
        ref_traj_(k, 2) = reference_trajectory[k](2);
    }
    */
    double amplitude = 2.0;   
    double frequency = 0.2;    
    double v_ref = 0.1;         

    for (int k = 0; k < N_; ++k) {
        double t = (k + 1) * dt_;
        double x_ref = x0_(0).scalar() + v_ref * t;
        double y_ref = amplitude * sin(frequency * x_ref);

        double dy_dx = atan2(y_ref - x0_(1).scalar(), x_ref - x0_(0).scalar());
        double theta_ref = atan2(sin(dy_dx), cos(dy_dx));

        ref_traj_(k, 0) = x_ref;
        ref_traj_(k, 1) = y_ref;
        ref_traj_(k, 2) = theta_ref;

        if (x_ref >= 5.0) {
            saveData();
            rclcpp::shutdown();
        }

        reference_trajectory_.push_back({x_ref, y_ref, 0.0});
    }
}



void ROS2Controllers::MPCController::saveData() {
    std::ofstream vehicle_file("/home/furkan/controller_ws/src/ros_controllers/results/result_mpc.csv");
    vehicle_file << "vehicle_x,vehicle_y,vehicle_z\n";
    for (const auto& pos : vehicle_positions_)
    {
        vehicle_file << pos[0] << "," << pos[1] << "," << pos[2] << "\n";
    }
    vehicle_file.close();

    std::ofstream ref_file("/home/furkan/controller_ws/src/ros_controllers/results/desired_mpc.csv");
    ref_file << "desired_x,desired_y,desired_z\n";
    for (const auto& ref : reference_trajectory_)
    {
        ref_file << ref[0] << "," << ref[1] << "," << ref[2] << "\n";
    }
    ref_file.close();
}



std::tuple<double, double, bool> ROS2Controllers::MPCController::computeControlSignal(
    const Eigen::VectorXd& state,
    const std::vector<Eigen::VectorXd>& reference_trajectory) {

    // Get the current state
    x0_ = DM::vertcat({state(0), state(1), state(2)});
    vehicle_positions_.push_back({state(0), state(1), 0.0});

    // Update trajectory
    updateReferenceTrajectory(reference_trajectory);

    // Update parameters
    DM p = SX::vertcat({x0_, SX::reshape(ref_traj_.T(), N_ * n_states_, 1)});
    DM x0 = SX::vertcat({SX::reshape(u0_, N_ * n_controls_, 1), SX::reshape(x_mpc_, (N_ + 1) * n_states_, 1)});
    DM lbg = DM::zeros(G_.size1(), 1);
    DM ubg = DM::zeros(G_.size1(), 1);

    std::map<std::string, DM> args;
    args["p"] = p;
    args["x0"] = x0;
    args["lbg"] = lbg;
    args["ubg"] = ubg;
    args["lbx"] = lbx_;
    args["ubx"] = ubx_;

    // Solve the problem
    auto sol = solver_(args);
    DM sol_x = sol.at("x");
    DM sol_u = sol_x(Slice(0, N_ * n_controls_));
    DM sol_states = sol_x(Slice(N_ * n_controls_, sol_x.size1()));

    sol_u = SX::reshape(sol_u, n_controls_, N_).T();
    sol_states = SX::reshape(sol_states, n_states_, N_ + 1).T();

    // Apply the first control input
    auto u_applied = sol_u(Slice(0, 1), Slice());
    double linear_velocity = static_cast<double>(u_applied(0, 0));
    double angular_velocity = static_cast<double>(u_applied(0, 1));

    // Update control and state predictions
    u0_ = DM::vertcat(std::vector<DM>{
        sol_u(Slice(1, N_), Slice()), 
        sol_u(Slice(N_ - 1, N_), Slice())
    });

    x_mpc_ = DM::vertcat(std::vector<DM>{
        sol_states(Slice(1, N_ + 1), Slice()), 
        sol_states(Slice(N_, N_ + 1), Slice())
    });

    // Calculating distance between vehicle and first reference point
    continous_linear_error_ = std::sqrt(std::pow(reference_trajectory[0](0) - state(0), 2) + 
        std::pow(reference_trajectory[0](1) - state(1), 2));

    double angle_error = std::abs(reference_trajectory[0](2) - state(2));
    
    std::cout << "Distance to first reference point: " << continous_linear_error_ << std::endl;
    std::cout << "Angle error: " << angle_error << std::endl;

    if (continous_linear_error_ < error_threshold_) {
        discrete_linear_error_ = continous_linear_error_;
        return std::make_tuple(linear_velocity, angular_velocity, true);
    }

    return std::make_tuple(linear_velocity, angular_velocity, false);
}

File: scripts/mpc.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import casadi as ca
import numpy as np
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
from tf_transformations import euler_from_quaternion

class MPCController(Node):
    def __init__(self):
        super().__init__('mpc_controller_node')

        # MPC parametreleri
        self.dt = 0.1       # Zaman adımı
        self.N = 20         # Öngörü ufku
        self.frequency = 10  # Kontrol döngüsü frekansı (Hz)

        # Durum ve kontrol değişkenleri
        x = ca.SX.sym('x')
        y = ca.SX.sym('y')
        theta = ca.SX.sym('theta')
        self.states = ca.vertcat(x, y, theta)
        n_states = self.states.size1()

        v = ca.SX.sym('v')
        omega = ca.SX.sym('omega')
        self.controls = ca.vertcat(v, omega)
        n_controls = self.controls.size1()

        # Sistem dinamiği (unicycle modeli)
        rhs = ca.vertcat(v * ca.cos(theta),
                         v * ca.sin(theta),
                         omega)

        # Dinamiği fonksiyon haline getirelim
        self.f = ca.Function('f', [self.states, self.controls], [rhs])

        # MPC değişkenleri
        self.U = ca.SX.sym('U', n_controls, self.N)          # Kontrol değişkenleri
        self.X = ca.SX.sym('X', n_states, self.N+1)          # Durum değişkenleri
        self.P = ca.SX.sym('P', n_states + self.N*n_states)  # Parametre vektörü

        # Amaç fonksiyonu ve kısıtlar
        obj = 0     # Amaç fonksiyonu
        g = []      # Kısıtlar listesi

        # Başlangıç durumu kısıtı
        st = self.X[:, 0]
        g.append(st - self.P[0:n_states])  # x0 = x_initial

        # Maliyet fonksiyonu (durum ve kontrol ağırlıkları)
        Q = ca.diagcat(20, 20, 1.0)        # Durum ağırlıkları
        R = ca.diagcat(0.5, 0.05)          # Kontrol ağırlıkları

        # MPC döngüsü
        for k in range(self.N):
            st = self.X[:, k]
            con = self.U[:, k]
            st_next = self.X[:, k+1]
            f_value = self.f(st, con)
            st_next_euler = st + self.dt * f_value
            g.append(st_next - st_next_euler)  # Sistem dinamiği kısıtı

            # Referans durumu
            ref = self.P[n_states + k*n_states : n_states + (k+1)*n_states]

            # Amaç fonksiyonu güncelleme
            obj = obj + ca.mtimes([(st - ref).T, Q, (st - ref)]) + ca.mtimes([con.T, R, con])

        # Değişkenleri tek bir vektörde birleştirelim
        OPT_variables = ca.vertcat(ca.reshape(self.U, -1, 1), ca.reshape(self.X, -1, 1))

        # Kısıtları birleştirelim
        g = ca.vertcat(*g)

        # NLP problemi tanımlayalım
        nlp_prob = {'f': obj, 'x': OPT_variables, 'g': g, 'p': self.P}

        # Çözümleyici ayarları
        opts = {'ipopt.max_iter': 2000,
                'ipopt.print_level': 0,
                'print_time': 0,
                'ipopt.acceptable_tol': 1e-8,
                'ipopt.acceptable_obj_change_tol': 1e-6}

        # Çözümleyiciyi oluştur
        self.solver = ca.nlpsol('solver', 'ipopt', nlp_prob, opts)

        # Kontrol ve durum tahminleri için başlangıç değerleri
        self.u0 = np.zeros((self.N, n_controls))            # Kontrol değişkenleri için başlangıç tahmini
        self.x_mpc = np.zeros((self.N+1, n_states))         # Durum değişkenleri için başlangıç tahmini

        # Durum ve kontrol sınırları
        self.lbx = np.full(OPT_variables.shape[0], -np.inf)
        self.ubx = np.full(OPT_variables.shape[0], np.inf)

        # Kontrol sınırları (TurtleBot3'ün fiziksel sınırlarına göre ayarlayın)
        v_max = 0.22  # m/s
        v_min = -v_max
        omega_max = 2.84  # rad/s
        omega_min = -omega_max

        for k in range(self.N):
            # Hız sınırları
            self.lbx[k*2] = v_min
            self.ubx[k*2] = v_max
            # Açısal hız sınırları
            self.lbx[k*2+1] = omega_min
            self.ubx[k*2+1] = omega_max

        # Mevcut durum
        self.x0 = np.zeros((n_states,))

        # Referans trajektori
        self.ref_traj = np.zeros((self.N, n_states))

        # ROS2 ayarları
        # Odometri aboneliği
        self.create_subscription(Odometry, '/odom', self.odom_callback, 10)

        # Kontrol komutu yayıncısı
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel', 10)

        self.position_ = Odometry()
        self.vehicle_positions_ = []
        self.reference_trajectory_ = []

        self.update_reference()


        # Kontrol döngüsünü başlat
        self.timer = self.create_timer(1.0 / self.frequency, self.control_loop)

    def odom_callback(self, msg):
        # Odometri verilerini al ve mevcut durumu güncelle
        position = msg.pose.pose.position
        self.position_ = msg
        orientation = msg.pose.pose.orientation
        quaternion = [orientation.x, orientation.y, orientation.z, orientation.w]
        _, _, yaw = euler_from_quaternion(quaternion)

        self.x0[0] = position.x
        self.x0[1] = position.y
        self.x0[2] = yaw

        self.vehicle_positions_.append((position.x, position.y, position.z))

    def control_loop(self):
        # Referans trajektoriyi güncelle
        self.update_reference()

        # Parametreleri güncelle
        args = {
            'p': np.concatenate([self.x0, self.ref_traj.reshape(-1)]),
            'x0': np.concatenate([self.u0.reshape(-1), self.x_mpc.reshape(-1)]),
            'lbg': np.zeros(self.X.shape[0]*(self.N+1)),
            'ubg': np.zeros(self.X.shape[0]*(self.N+1)),
            'lbx': self.lbx,
            'ubx': self.ubx
        }

        # Problemi çöz
        sol = self.solver(**args)
        sol_u = sol['x'][0:self.N*2].full().reshape(self.N, 2)
        sol_x = sol['x'][self.N*2:].full().reshape(self.N+1, 3)

        # İlk kontrol girdisini uygula
        u_applied = sol_u[0, :]
        cmd = Twist()
        cmd.linear.x = float(u_applied[0])
        cmd.angular.z = float(u_applied[1])
        self.cmd_pub.publish(cmd)

        # Kontrol ve durum tahminlerini güncelle
        self.u0 = np.vstack((sol_u[1:, :], sol_u[-1, :]))
        self.x_mpc = np.vstack((sol_x[1:, :], sol_x[-1, :]))

    def update_reference(self):
        # Sinüs dalga referans trajektorisi
        amplitude = 1.0     # Sinüs dalgasının genliği
        frequency = 0.5     # Sinüs dalgasının frekansı
        v_ref = 0.1         # x ekseninde referans hız

        if self.position_.pose.pose.position.x >= 5.0:
            print("Hedefe ulaşıldı!")
            self.save_data()
            return 

        for k in range(self.N):
            t = (k + 1) * self.dt  # Gelecekteki zaman adımı
            x_ref = self.x0[0] + v_ref * t
            y_ref = amplitude * np.sin(frequency * x_ref)
            
            # y'nin x'e göre türevi (dy/dx)
            dy_dx = amplitude * frequency * np.cos(frequency * x_ref)
            theta_ref = np.arctan2(dy_dx, 1)  # dx/dx = 1 olduğu varsayımıyla

            self.ref_traj[k, 0] = x_ref
            self.ref_traj[k, 1] = y_ref
            self.ref_traj[k, 2] = theta_ref

        for ref_state in self.ref_traj:
            self.reference_trajectory_.append((ref_state[0], ref_state[1], 0.0)) 



    def save_data(self):
        with open('/home/furkan/controller_ws/src/ros_controllers/results/result_mpc.csv', 'w') as f:
            f.write('vehicle_x,vehicle_y,vehicle_z\n')
            for pos in self.vehicle_positions_:
                f.write(f'{pos[0]},{pos[1]},{pos[2]}\n')

        with open('/home/furkan/controller_ws/src/ros_controllers/results/desired_mpc.csv', 'w') as f:
            f.write('desired_x,desired_y,desired_z\n')
            for ref in self.reference_trajectory_:
                f.write(f'{ref[0]},{ref[1]},{ref[2]}\n')


def main(args=None):
    rclpy.init(args=args)
    mpc_controller = MPCController()
    rclpy.spin(mpc_controller)
    mpc_controller.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
